src/Estimators/NESpaceGrid.cpp

```diff
@@ -588,16 +588,31 @@ void NESpaceGrid<REAL>::accumulate(const ParticlePos& R,
       {
         particles_outside[p] = false;
         Point u              = dot(axinv_, (R[p] - origin_));
+	// numerical accuracy of the dot is not good enough for this to be not produce out of bounds indexes
+	// with REAL = float, u 10^-8 outside of umin_ to  umax_ is not uncommon i.e  p > (1 in 32000)
+	// for the coordinate ranges typical of our simulations. I guess the branch predictor will mostly
+	// mitigate this.
+	auto gmapIndex = [this](int d, const auto& u) { int raw_index = floor((u[d] - this->umin_[d]) * this->odu_[d]);
+	  if ( raw_index == gmap_[d].size() )
```

Comment:
Thanks for diagnosing this Peter. A classic case of someone mapping a floating point number into an integer index and not taking care to ensure that the resultant integer is within range given the issues for floating point tolerances, rounding etc....

In the code here, are you certain that the mapping is correct for periodic boundary conditions cases? if the case of arriving at a negative index we could map to either the zeroth or nth-1 grid point. Here you are mapping -1 to zero (first choice). Should we instead be adding/removing gmap_[d].size to get within range 0-gmap_[d].size-1? This is my inclination. This would also address Ye's question, if we haven't already ensured that the [d] - this->umin_[d] is positive.

Reply:
So you could also claim that for nth grid point you can map to zeroth or nth-1, my observation is misses on both sides occurred.

And
```
std::floor(u[d] * odu[d]) - std::floor(umin[d] * odu[d])
```
seems like it needs a different anlysis umin now could always be truncated even if was arbitrarily close to the next integer but below. Whereas u would still be free to be -1 or +1 about the range. I guess then -2 goes to nth-1 and -1 goes to 0;

There are still missing constraints at the floating point level on the grids but that is a lack on input validation which has always been there, any of these schemes is likely to catch them since almost certainly indexes off by more than one will occur.